google-cloud-firestore/src/main/java/com/google/cloud/firestore/telemetry/MetricsUtil.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore.telemetry;
+
+import static com.google.cloud.firestore.telemetry.BuiltinMetricsConstants.*;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.api.gax.tracing.ApiTracerFactory;
+import com.google.api.gax.tracing.MetricsTracerFactory;
+import com.google.api.gax.tracing.OpenTelemetryMetricsRecorder;
+import com.google.cloud.firestore.FirestoreException;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.Status;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.metrics.DoubleHistogram;
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * A utility interface for trace collection. Classes that implement this interface may make their
+ * own design choices for how they approach trace collection. For instance, they may be no-op, or
+ * they may use a particular tracing framework such as OpenTelemetry.
+ */
+public class MetricsUtil {
+
+  private final FirestoreOptions firestoreOptions;
+  private OpenTelemetry openTelemetry;
+  private ApiTracerFactory otelApiTracerFactory;
+
+  private Meter meter;
+  private DoubleHistogram endToEndRequestLatency;
+  private DoubleHistogram firstResponseLatency;
```

Comment:
This puts otel code on the main execution path of the Firestore sdk, which is something I would like to avoid. Using enabled/disabled MetricsUtil can resolve that.